src/Arguments.php

```diff
@@ -0,0 +1,216 @@
+     *
+     * In case $value is left empty, the argument will be treated as a boolean option. If the
+     * name and value are not separated by a space, pass the correct separator as a third
+     * parameter.
+     *
+     * Unless an option is registered (in this class) as an application option, we will assume
+     * the option is meant to be passed through to PHPUnit.
+     *
+     * @param  string $name
+     * @param  mixed $value
+     * @param  string $separator
+     */
+    public function addArgument($name, $value = null, $separator = '=')
+    {
+        $valueWithSeparator = is_null($value) ? $value : ['value' => $value, 'separator' => $separator];
+
+        if ($this->isApplicationOption($this->optionName($name))) {
+            $this->applicationOptions[$name] = $valueWithSeparator;
+
+            return;
+        }
+
+        $this->phpUnitOptions[$name] = $valueWithSeparator;
+    }
+
+    /**
+     * Returns a string containing all PHPUnit options with corresponding values in the correct format.
+     *
+     * @return string
+     */
+    public function phpUnitArguments()
+    {
+        $arguments = [];
+
+        foreach ($this->phpUnitOptions as $name => $optionValue) {
+            if (is_null($optionValue)) {
+                $arguments[] = $name;
+                continue;
+            }
+            $arguments[] = $name.$optionValue['separator'].$optionValue['value'];
+        }
+
+        if (! empty($this->testFile)) {
+            $arguments[] = $this->testFile;
+        }
+
+        return implode(' ', $arguments);
+    }
+
+    /**
+     * Returns an array representation containing application options, PHPUnit options and the test file.
+     *
+     * @return array
+     */
+    public function toArray()
+    {
+        return array_merge(
+            $this->applicationOptions,
+            $this->phpUnitOptions,
+            empty($this->testFile) ? [] : [$this->testFile => null]
+        );
+    }
+
+    /**
+     * Parses a raw arguments string.
+     *
+     * The following types of arguments will be parsed: space separated arguments, key/value and boolean arguments,
+     * and test file name arguments. Space separated arguments are key/value arguments separated by a space. Boolean
+     * and key/value arguments are regular options that either take no parameters or take a =-separated parameter.
+     * Test file name arguments are file names (without option flags).
+     *
+     * @param $argumentsInput
+     * @return $this
+     */
+    protected function parse($argumentsInput)
+    {
+        $arguments = explode(' ', $argumentsInput);
+
+        // Keeps track of option name belonging to value when option name and value are space separated
+        $nextArgumentBelongsTo = false;
+
+        // PHPUnit only uses first file when multiple are given
+        $fileSet = false;
+
+        foreach ($arguments as $argument) {
+            if ($nextArgumentBelongsTo) {
```

Comment:
The `$filterArgumentHasBeenUsed` and `$nextArgumentBelongsTo` variables unfortunately make it hard do this. I didn't like it very much either, but I couldn't come up with anything simpeler than this. The only other option, but I think it's a bit overkill, is promoting the 'parser' to its own class. The two variables can then become class fields, which make it possible to go with early returns and stuff.